libsqlite3-sys/src/lib.rs

```diff
@@ -49,6 +49,26 @@ pub enum Limit {
     SQLITE_LIMIT_WORKER_THREADS = 11,
 }
 
+#[cfg(feature = "loadable_extension_embedded")]
+// bindings were built with loadable_extension_embedded:
+// define sqlite3_api as an extern since this code will be embedded
+// within a loadable extension that defines and exports this itself
+extern "C" {
+    #[no_mangle]
+    pub static mut sqlite3_api: *mut sqlite3_api_routines;
```

Comment:
This worries me. I'd feel better if we synchronized access to it somehow...

At a minimum I guess we should document that this may be read from without synchronization by any call into rusqlite, and so the safety contract for use is that it must be initialized strictly before any thread could possibly read from it.

This goes for both declarations of `sqlite3_api`.

Reply:
The `sqlite3_api` symbol is the interface provided by sqlite to its run-time loadable extensions. The wrapper functions we generate (see https://github.com/rusqlite/rusqlite/blob/cf658af142399be749bacb0387d54ce836b85f1a/libsqlite3-sys/build.rs#L635) do a test for it not being null. 

When we are compiled as an embedded extension, the contract with the extension we are embedded within is that the `sqlite3_api` symbol is initialised before our entrypoint is called. 

When we are a non-embedded extension, it is our responsibility to set `sqlite3_api` as the first thing we do within our extension entrypoint (e.g. https://github.com/rusqlite/rusqlite/blob/cf658af142399be749bacb0387d54ce836b85f1a/dummy-extension/src/lib.rs#L19). Even so, we initialise the pointer to null in the declaration and the wrapper function protection should protect us until our entrypoint sets it.

I suppose there is no reason to expose the `sqlite3_api` symbol directly outside of libsqlite3-sys -- the only appropriate direct interactions with it are from the generated wrapper functions and to set it (once) at the start of the extension entrypoint. 

Rather than asking the entrypoint to set the symbol directly, we could expose a pub function to initialise the extension, in which case the symbol could be private to the libsqlite3-sys crate, and I suppose we could also make sure that it is only initialised once. 

I.e. something like:
```
use std::sync::Once;
static SQLITE_EXTENSION_INIT: Once = Once::new();

#[cfg(feature = "loadable_extension_embedded")]
pub fn sqlite_extension_init(_p_api: *mut ffi::sqlite3_api_routines) {}

#[cfg(all(
    feature = "loadable_extension",
    not(feature = "loadable_extension_embedded")
))]
pub fn sqlite_extension_init(p_api: *mut ffi::sqlite3_api_routines) {
    SQLITE_EXTENSION_INIT.call_once(|| {
        unsafe {
            ffi::sqlite3_api = p_api;
        }
    });
}
```

Would that alleviate some of your concerns?

Reply:
More or less. Of course, `sqlite_extension_init` should be an `unsafe fn`.

Reply:
Discussion moved to https://github.com/rusqlite/rusqlite/pull/910#discussion_r765315952